R/ETagMiddleware.R

```diff
@@ -0,0 +1,232 @@
+#' app$process_request(req)
+#'
+#' # provide both headers: If-None-Match takes precedence
+#' # in this case:
+#' #  - if none match => modified (No cache)
+#' #  - if modified since => NOT MODIFIED (cached)
+#' # => Overall: modified = no cache
+#' modified_since = format(last_modified + 1, "%FT%TZ")
+#' req = Request$new(path = "/example.txt",
+#'                   headers = list("If-None-Match" = "CLEARLY WRONG",
+#'                                  "If-Modified-Since" = modified_since))
+#' app$process_request(req)
+#'
+#' #############################################################################
+#'
+#' # use an alternative hash function (use name of the file)
+#' hash_on_filename = function(x) x
+#' # also use an alternate last_modified time function
+#' always_1900 = function(x) as.POSIXlt("1900-01-01 12:34:56", tz = "GMT")
+#'
+#' app = Application$new(middleware = list(
+#'   ETagMiddleware$new(hash_function = hash_on_filename,
+#'                      last_modified_function = always_1900)
+#' ))
+#' app$add_static(path = "/", file_path = static_dir)
+#'
+#' req = Request$new(path = "/example.txt")
+#' (res = app$process_request(req))
+#'
+#' filename = res$body[["file"]]
+#' req = Request$new(path = "/example.txt",
+#'                   headers = list("If-None-Match" = filename))
+#' app$process_request(req)
+ETagMiddleware = R6::R6Class(
+  classname = "EtagMiddleware",
+  inherit = Middleware,
+  public = list(
+    #' @field hash_function Function that takes an object or file and computes
+    #' the hash of it
+    hash_function = NA,
+    #' @field last_modified_function Function that takes an object or file and
+    #' computes the last time it was modified
+    last_modified_function = NA,
+    #' @description
+    #' Creates ETag middleware object
+    #' @param routes Routes paths to protect.
+    #' @param match How routes will be matched: exact or partial (as prefix).
+    #' @param id Middleware id.
+    #' @param hash_function a function that generates the ETag hash.
+    #' The function takes the body of the response and returns a single
+    #' character. Default is crc32 using [digest::digest].
+    #' @param last_modified_function a function that takes the body of the
+    #' response and returns the last time this was changed. The default is to
+    #' take the mtime (last time the file was modified) if its a file,
+    #' if the body does not contain a file, the current time is returned (
+    #' resulting in no caching)
+    initialize = function(routes = "/", match = "partial",
+                          id = "ETagMiddleware",
+                          hash_function = function(body) {
+                            if ("file" %in% names(body)) {
+                              digest::digest(file = body[["file"]],
+                                             algo = "crc32")
+                            } else {
+                              digest::digest(body, algo = "crc32")
```

Comment:
I'm thinking if we should check if `body` is already `raw` or `character` and avoid serialization. This will allow to get a small edge on performance. @artemklevtsov  @DavZim thoughts?

Reply:
You mean replace line 133 with something along the lines of:

```r
if (any(c("character", "raw") %in% class(body)) {
  digest::digest(body, serialize = FALSE, algo = "crc32")
} else {
  digest::digest(body, algo = "crc32")
}
```
For an 11MB raw vector (of some random file I used), it reduced the computation time by 35% from 20ms down to 13ms.

At the same time, the user or whoever uses or implements the originator of the request and computes the If-Not-Match header of the ETag needs to be able to implement the same logic to get the same hash.

Just to check my understanding, this would return a different hash for an object, correct? So the difference in hashes is expected, correct?

``` r
body = as.raw(c(123, 123, 123))

body
#> [1] 7b 7b 7b
serialize(body, connection = NULL)
#>  [1] 58 0a 00 00 00 03 00 04 01 02 00 03 05 00 00 00 00 06 43 50 31 32 35 32 00
#> [26] 00 00 18 00 00 00 03 7b 7b 7b
# the way digest::digest calls serialize internally
serialize(body, connection = NULL, ascii = FALSE, version = 2)
#>  [1] 58 0a 00 00 00 02 00 04 01 02 00 02 03 00 00 00 00 18 00 00 00 03 7b 7b 7b

digest::digest(body)
#> [1] "f64bde36e6d414f445f7b544334b25dc"
digest::digest(body, serialize = FALSE)
#> [1] "317823dbcada605031208cc5d8159c59"
# the way digest::digest calls serialize internally
digest::digest(serialize(body, connection = NULL, ascii = FALSE, version = 2),
               serialize = FALSE)
#> [1] "515bff8a9cdf0f5c6fc8f42387406c05"
```

I am not sure if this speedup is worth the confusion and added complexity. If we go with this route, we could think of creating the hash and last-modified functions as first-citizen functions and exporting them to the userland as well, so the user can use them outside as well (with the added benefit of making tests more clear).

Reply:
yeah, I think it doesn't worth